### ISSUE TITLE
Presence of record in src/test/java causes error during execution of JUnit 4

### ISSUE THREAD

**Issue by rweisleder** [gradle/gradle#14714]
I created a simple Java project with Gradle which contains a JUnit 4 test class and a record (Java 14/15 preview feature), both in `src/test/java`:
```java
public record Person(String name) {
}
```
```java
import org.junit.Assert;
import org.junit.Test;

public class PersonTest {

    @Test
    public void test_getName() {
        Person person = new Person("John Doe");
        Assert.assertEquals("John Doe", person.name());
    }
}
```

### Expected Behavior

Running `./gradlew test` should only execute `PersonTest` as test class and succeed.

### Current Behavior

Running `./gradlew test` causes JUnit 4.13 trying to execute `Person` as test class. This attempt fails with:
```
org.example.Person > initializationError FAILED
    org.junit.runners.model.InvalidTestClassError: Invalid test class 'org.example.Person':
      1. Test class should have exactly one public zero-argument constructor
      2. No runnable methods
        at org.junit.runners.ParentRunner.validate(ParentRunner.java:525)
        at org.junit.runners.ParentRunner.<init>(ParentRunner.java:102)
        at org.junit.runners.BlockJUnit4ClassRunner.<init>(BlockJUnit4ClassRunner.java:84)
        at org.junit.runners.JUnit4.<init>(JUnit4.java:23)
        at org.junit.internal.builders.JUnit4Builder.runnerForClass(JUnit4Builder.java:10)
        at org.junit.runners.model.RunnerBuilder.safeRunnerForClass(RunnerBuilder.java:70)
        at org.junit.internal.builders.AllDefaultPossibilitiesBuilder.runnerForClass(AllDefaultPossibilitiesBuilder.java:37)
        at org.junit.runners.model.RunnerBuilder.safeRunnerForClass(RunnerBuilder.java:70)
        at org.junit.internal.requests.ClassRequest.createRunner(ClassRequest.java:28)
        at org.junit.internal.requests.MemoizingRequest.getRunner(MemoizingRequest.java:19)
        at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.runTestClass(JUnitTestClassExecutor.java:78)
        at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:58)
        at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:38)
        at org.gradle.api.internal.tasks.testing.junit.AbstractJUnitTestClassProcessor.processTestClass(AbstractJUnitTestClassProcessor.java:62)
        at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:51)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:564)
        at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
        at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
        at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
        at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
        at com.sun.proxy.$Proxy2.processTestClass(Unknown Source)
        at org.gradle.api.internal.tasks.testing.worker.TestWorker.processTestClass(TestWorker.java:119)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:564)
        at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
        at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
        at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:182)
        at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:164)
        at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:414)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
        at java.base/java.lang.Thread.run(Thread.java:832)
```

This exception is logged at debug level:
```
[DEBUG] [org.gradle.api.internal.tasks.testing.detection.AbstractTestFrameworkDetector] Failed to read class file C:\workspace\spielwiese\test-records-with-gradle-and-junit\build\classes\java\test\org\example\Person.class; assuming it's a test class and continuing
java.lang.UnsupportedOperationException: This feature requires ASM8_EXPERIMENTAL
        at org.objectweb.asm.ClassVisitor.visitRecordComponentExperimental(ClassVisitor.java:304)
        at org.objectweb.asm.ClassReader.readRecordComponent(ClassReader.java:929)
        at org.objectweb.asm.ClassReader.accept(ClassReader.java:704)
        at org.objectweb.asm.ClassReader.accept(ClassReader.java:401)
        at org.gradle.api.internal.tasks.testing.detection.AbstractTestFrameworkDetector.readClassFile(AbstractTestFrameworkDetector.java:127)
        at org.gradle.api.internal.tasks.testing.detection.AbstractTestFrameworkDetector.processTestClass(AbstractTestFrameworkDetector.java:153)
        at org.gradle.api.internal.tasks.testing.detection.AbstractTestFrameworkDetector.processTestClass(AbstractTestFrameworkDetector.java:139)
        at org.gradle.api.internal.tasks.testing.detection.DefaultTestClassScanner$1.visitClassFile(DefaultTestClassScanner.java:60)
        at org.gradle.api.internal.tasks.testing.detection.DefaultTestClassScanner$ClassFileVisitor.visitFile(DefaultTestClassScanner.java:79)
        at org.gradle.api.internal.file.collections.DefaultDirectoryWalker$PathVisitor.visitFile(DefaultDirectoryWalker.java:115)
        at org.gradle.api.internal.file.collections.DefaultDirectoryWalker$PathVisitor.visitFile(DefaultDirectoryWalker.java:68)
        at java.base/java.nio.file.Files.walkFileTree(Files.java:2804)
        at org.gradle.api.internal.file.collections.DefaultDirectoryWalker.walkDir(DefaultDirectoryWalker.java:62)
        at org.gradle.api.internal.file.collections.DirectoryFileTree.walkDir(DirectoryFileTree.java:149)
        at org.gradle.api.internal.file.collections.DirectoryFileTree.visitFrom(DirectoryFileTree.java:127)
        at org.gradle.api.internal.file.collections.DirectoryFileTree.visit(DirectoryFileTree.java:112)
        at org.gradle.api.internal.file.collections.FileTreeAdapter.visit(FileTreeAdapter.java:87)
        at org.gradle.api.internal.file.CompositeFileTree.visit(CompositeFileTree.java:102)
        at org.gradle.api.internal.tasks.testing.detection.DefaultTestClassScanner.detectionScan(DefaultTestClassScanner.java:57)
        at org.gradle.api.internal.tasks.testing.detection.DefaultTestClassScanner.run(DefaultTestClassScanner.java:51)
        at org.gradle.api.internal.tasks.testing.processors.TestMainAction.run(TestMainAction.java:55)
        at org.gradle.api.internal.tasks.testing.detection.DefaultTestExecuter.execute(DefaultTestExecuter.java:121)
        at org.gradle.api.internal.tasks.testing.detection.DefaultTestExecuter.execute(DefaultTestExecuter.java:52)
        at org.gradle.api.tasks.testing.AbstractTestTask.executeTests(AbstractTestTask.java:484)
        at org.gradle.api.tasks.testing.Test.executeTests(Test.java:664)
...
```

All other tests in my project get executed normally.

### Context

While implementing https://github.com/TNG/ArchUnit/issues/295 I tried to add a test to verify that ArchUnit imports records correctly. Therefore I added a "record under test" in src/test/java. After adding the record, the execution started to fail.

### Steps to Reproduce

Checkout [rweisleder/test-records-with-gradle-and-junit](https://github.com/rweisleder/test-records-with-gradle-and-junit) and run `./gradlew test`.

### Environment

This issue appeared with Gradle 6.6.1 (and probably also with older versions) and is also reproducible with Gradle 6.7 RC3.

### Workaround

My current workaround is to explicitly exclude the "record under test" in the build.gradle.
```
test {
    exclude 'org/example/Person.class'
}
```

**Comment by ljacomet** [comment]
This requires full support of `Record` in Gradle, which are only feature preview in Java 15.
Gradle will need to support that when `Record` are a released language feature.

**Comment by jjohannes** [comment]
If _Records_ become a stable feature in Java 16, this should be fixed with official Java 16 support in Gradle (#13481).

**Comment by mikehearn** [comment]
Is it valid to assume any unparseable class file is a test? If that logic was inverted people would be able to use new Java features in non-test classes, which would be useful.

**Comment by sbrannen** [comment]
> Is it valid to assume any unparseable class file is a test? 

I also wonder if that's the best approach for Gradle to take.

@ljacomet, @jjohannes, @marcphilipp, thoughts?

**Comment by marcphilipp** [comment]
> > Is it valid to assume any unparseable class file is a test?
> 
> I also wonder if that's the best approach for Gradle to take.

It's certainly safer so unparseable test classes don't go unnoticed.

@ljacomet Could we introduce a temporary toggle to invert that behavior or opt-in to experimental ASM features?

**Comment by mikehearn** [comment]
I'm not sure it's a safety feature given that the error message given is misleading - the code clearly doesn't expect this which is why it's only logged at debug level.

If you want a bit of magic that yields a better error message, you could catch ASM exceptions and simply print a warning, or make it an error if the source file contains a line starting with `import org.junit.` for example.

**Comment by cocorossello** [comment]
I just realized that incremental compilation is also broken if a record is present (tested in jdk15)

**Comment by rweisleder** [comment]
This issue seems to be fixed in 7.0 RC2.

I updated my demo project and it works both with Java 14 (`--enable-preview`) and with Java 16.

**Comment by rweisleder** [comment]
It also works with the freshly released version 7.0.

Thanks a lot for your efforts!
